File: org.summer.view.widget/src/org/summer/view/widget/internal/ExplicitObjectRef.java
package org.summer.view.widget.internal;

import org.summer.view.widget.DependencyObject;
import org.summer.view.widget.WeakReference;

//#region ExplicitObjectRef
/// <summary> Explicit Object reference. </summary>
/*internal*/public final class ExplicitObjectRef extends ObjectRef 
{
    //----------------------------------------------------- 
    // 
    //  Constructors
    // 
    //-----------------------------------------------------

    /// <summary> Constructor. </summary>
    /*internal*/public ExplicitObjectRef(Object o) 
    {
        if (o instanceof DependencyObject) 
            _element = new WeakReference(o); 
        else
            _object = o; 
    }

    //------------------------------------------------------
    // 
    //  Public Methods
    // 
    //----------------------------------------------------- 

    /// <summary> Returns the referenced Object. </summary> 
    /// <param name="d">Element defining context for the reference. </param>
    /// <param name="args">See ObjectRefArgs </param>
    /*internal*/ /*override*/ Object GetObject(DependencyObject d, ObjectRefArgs args)
    { 
        return (_element != null) ? _element.Target : _object;
    } 

    /// <summary>
    /// true if the ObjectRef uses the mentor of the target element, 
    /// rather than the target element itself.
    /// </summary>
    protected /*override*/ boolean ProtectedUsesMentor
    { 
        get { return false; }
    } 

    /*internal*/ /*override*/ String Identify()
    { 
        return "Source";
    }

    //------------------------------------------------------ 
    //
    //  Private Fields 
    // 
    //------------------------------------------------------

    Object _object;
    WeakReference _element; // to DependencyObject (bug 986435)
}

File: summer/WebContent/js/MS/internal/data/ExplicitObjectRef.js
/**
 * Second Check 12-27
 * ExplicitObjectRef
 */
define(["dojo/_base/declare", "system/Type", "internal.data/ObjectRef"], 
		function(declare, Type, ObjectRef){
	var ExplicitObjectRef = declare("ExplicitObjectRef", ObjectRef, {
		constructor:function( o ){
			this._object = o; 
		},

        /// <summary> Returns the referenced object. </summary> 
        /// <param name="d">Element defining context for the reference. </param>
        /// <param name="args">See ObjectRefArgs </param>
        /*internal override object */
		GetObject:function(/*DependencyObject*/ d, /*ObjectRefArgs*/ args)
        { 
            return this._object;
        }, 
 
        /*internal override string */
        Identify:function()
        { 
            return "Source";
        }
	});
	
	Object.defineProperties(ExplicitObjectRef.prototype,{
        /// <summary>
        /// true if the ObjectRef uses the mentor of the target element, 
        /// rather than the target element itself.
        /// </summary>
        /*protected override bool*/ 
		ProtectedUsesMentor:
        { 
            get:function() { return false; }
        } 
	});
	
	ExplicitObjectRef.Type = new Type("ExplicitObjectRef", ExplicitObjectRef, [ObjectRef.Type]);
	return ExplicitObjectRef;
});
